### PR TITLE
[WIP] Change cursor style when a queryable layer is available

### LINKED ISSUE #878: 6.2.15 Tooltip DESKTOP : change mousepointer on mousover  on features with tooltip
ALL OS, All DESKTOP Browsers

When a feature of an active layer has a tooltip, change mousepointer on mousover  on features with tooltip (from arrow to a questionmark) ( I know, UTFGrid we dont' have have and won't do)

### PR DESCRIPTION
**Pull request by oterral**
Fix #878

[Test](http://mf-geoadmin3.dev.bgdi.ch/teo_cursor/?lang=fr&topic=luftbilder&bgLayer=ch.swisstopo.pixelkarte-grau&X=121948.74&Y=647906.61&zoom=4&layers=ch.bazl.luftfahrthindernis,ch.swisstopo.lubis-luftbilder_schwarzweiss,ch.swisstopo.lubis-luftbilder_farbe&catalogNodes=1179,1180,1186&layers_timestamp=,99991231,99991231)

### CODE REVIEW

**Comment by ponceta** [comment]
Nice, this is great! A little bit laaaaaggy on IE 11 but works well on chrome and FF

**Comment by loicgasser** [comment]
Yes this is cool. I would not add it on mobile.

**Comment by oterral** [comment]
I'ts already not added to mobile

**Comment by oterral** [comment]
@ponceta why laggy it works the same on our IE11?

**Comment by gjn** [comment]
Killer. This is a killer!

File: src/components/map/MapService.js [hunk 1]
```diff
@@ -765,7 +765,8 @@
                     layer.minResolution),
                 tileLoadFunction: tileLoadFunction,
                 url: getWmtsGetTileUrl(layer.serverLayerName,
-                  layer.format)
+                  layer.format),
+                crossOrigin: 'anonymous'
```
Comment (gjn) [inline src/components/map/MapService.js]: Why is this needed?

**Comment by gjn** [comment]
LGTM: Big fan of this.

Why still WIP?

**Comment by ponceta** [comment]
@oterral try http://mf-geoadmin3.dev.bgdi.ch/teo_cursor/?lang=fr&topic=vu&bgLayer=ch.swisstopo.pixelkarte-grau&X=154820.67&Y=522532.85&zoom=4&catalogNodes=1318&layers=ch.astra.unfaelle-personenschaeden_alle

for example. Works fine on chrome and FF but not on IE 11

It is anyway a great improvement!

**Comment by gjn** [comment]
@davidoesch @cedricmoullet This is a very cool feature. But it has some drawbacks on User interaction that is best visible on the luftfahrthindernisse layers [1]. It's mainly about layers with point features that have quite large symbols. The client detects that a layer is drawing something (the symbol on the map), but our back-end is unaware of the symbol and only check for a point. So a click will not result in a selected object, despite having the hand-cursor.

IMO, this needs to be adressed first.

Thoughts?

[1] http://mf-geoadmin3.dev.bgdi.ch/teo_cursor/?lang=fr&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&X=154820.67&Y=522532.85&zoom=4&catalogNodes=1379,1381,1514&layers=ch.bazl.luftfahrthindernis

**Comment by AFoletti** [comment]
is there a way to force the detection of an exact RGB value and ignore the others? If yes, we could use an "invisible" STYLE (such as https://github.com/geoadmin/web-emapis/issues/55) to force detection on the geometry and not the symbol.

Polygons: not a big problem, but implement the invisible style where simbology does not fill the poly (hatches, outlines only)
Lines: same as polygons (dashed lines)
Points: the invisible style will be a circle on the geometry, triggering the cursor change (we have to adapt the size to the tooltip-click buffer)

Even if this is possible, I still see a potential problem with CH-covering polygons, where the cursor will always be a hand. Correct from the behavior point, but maybe not so nice to see

**Comment by davidoesch** [comment]
How does Google solve that? AFAIK change pointer on hover is only working for points/line there ( since the do not have polygons).




Having the mouse icon changed on poly : TBD. But we have to be consistent for point line and poly 




First : solving the issue of hand icon an click. I hate to say it but utfgrid would solve this.

File: src/components/map/MapService.js (hunk 1 above)
Comment (oterral) [inline src/components/map/MapService.js]: The hit detection creates a canvas then try to export the data, and because of CORS you must set the crossOrigin to anonymous to allow the export.

**Comment by cedricmoullet** [comment]
80/20 -> I'm in favour to merge it. 
@davidoesch ok ?

**Comment by davidoesch** [comment]
Ok

**Comment by AFoletti** [comment]
I'm still no a big fan of the actual state of the whole. It's actually based on the SYMBOL, while the actual identify result is based on the GEOMETRY. The two are not always the same.
We already saw how map.geo.admin.ch is complicated for the standard user... I think this is really a great feature if done right, but only adds confusion if partially solved

**Comment by gjn** [comment]
Not a fan of the exceptions either. Question is, how many layers are affected by this?

**Comment by davidoesch** [comment]
Agree Keep IT simple. Google maps only has to deal with points... Other thoughts ideas?

**Comment by AFoletti** [comment]
I can't quantify exactly how many layers, but each hollow (borders layers) or hatched polygon layer is a problem, not to mention the luftfahrthindernisse or other strangely-symbolized point features. Even a big round point is a problem, because the symbol will be much bigger than the actual IDENTIFY buffer (treasurehunt anyone?). Or a hollow point for that matter...

All in all, @cedricmoullet is quite right. This is really 80/20. I would say we have real problems in 20-30 layers and the rest works perfectly

**Comment by davidoesch** [comment]
Consistency goes a long way. Explaining the case for 30 layers - costs are 2 high

**Comment by gjn** [comment]
Here are, I think, most use cases.

Good -> Typical ploygon layer: http://mf-geoadmin3.dev.bgdi.ch/?X=173770.58&Y=669179.42&zoom=1&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.swisstopo-vd.ortschaftenverzeichnis_plz&layers_opacity=0.75

Not ideal -> Typical polyline layer: http://mf-geoadmin3.dev.bgdi.ch/?X=173770.58&Y=669179.42&zoom=1&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.swisstopo.swissboundaries3d-gemeinde-flaeche.fill

Quite good -> Typical point layer?: http://mf-geoadmin3.dev.bgdi.ch/?X=173770.58&Y=669179.42&zoom=1&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.astra.unfaelle-personenschaeden_alle

Could be misleading -> Symbolic point layer (as described by Amrogio, circle too big): http://mf-geoadmin3.dev.bgdi.ch/?X=116470.58&Y=657929.42&zoom=4&lang=de&topic=funksender&bgLayer=ch.swisstopo.pixelkarte-farbe&catalogNodes=403,408&layers=ch.bakom.radio-fernsehsender,ch.bakom.mobil-antennenstandorte-gsm,ch.bakom.mobil-antennenstandorte-umts,ch.bakom.mobil-antennenstandorte-lte

Then there's luftfahrhindernisse layer, which is misleading for points: http://mf-geoadmin3.dev.bgdi.ch/?X=161412.50&Y=605302.50&zoom=8&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.bazl.luftfahrthindernis

I can't even say what's worse, false positives or false negatives ;=)

**Comment by gjn** [comment]
Also, there are other ways to show to the user that he can acutally click the map to get information....like hints, tip of the day, etc. Not sure how they would fit into the usability discussion we also have.

**Comment by davidoesch** [comment]
Tested the links above: and IMHO there is a *minor* issue for the user  for

A) Not ideal -> Typical polyline layer: http://mf-geoadmin3.dev.bgdi.ch/?X=173770.58&Y=669179.42&zoom=1&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.swisstopo.swissboundaries3d-gemeinde-flaeche.fill 
->  this polyline layer is actually a polygon layer (according to the highlight geometry...) so the cursor style shoudl be dependent where we have a highlight geometry

B)  luftfahrhindernisse layer, which is misleading for points: http://mf-geoadmin3.dev.bgdi.ch/?X=161412.50&Y=605302.50&zoom=8&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.bazl.luftfahrthindernis

-> issue exists already without mouse icon change: if you click on the top north part of the single obstacle , there is a loading icon showing up but nothing happens. so as well here: show the mouse over icon change based on the highlight geometry

-> IMHO: if we use in general the highlight geometry as base for the mouse icon change usability would be nice. (but this will have a serious load on the back-end I assume)

**Comment by cedricmoullet** [comment]
I still think that we can apply here the 80/20 rule. As David says, the
problem is already here.
What about having a bigger tolerance in order to avoid false negative ?

And in 80% of the case, this uses a design pattern that is known for the
user: "If the cursor is a hand, I can click on it ¨

On Sat, Mar 7, 2015 at 11:28 AM, davidoesch <notifications@github.com>
wrote:

> Tested the links above: and IMHO there is a *minor* issue for the user for
>
> A) Not ideal -> Typical polyline layer:
> http://mf-geoadmin3.dev.bgdi.ch/?X=173770.58&Y=669179.42&zoom=1&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.swisstopo.swissboundaries3d-gemeinde-flaeche.fill
> -> this polyline layer is actually a polygon layer (according to the
> highlight geometry...) so the cursor style shoudl be dependent where we
> have a highlight geometry
>
> B) luftfahrhindernisse layer, which is misleading for points:
> http://mf-geoadmin3.dev.bgdi.ch/?X=161412.50&Y=605302.50&zoom=8&lang=de&topic=aviation&bgLayer=ch.swisstopo.pixelkarte-grau&catalogNodes=1379,1381,1514&layers=ch.bazl.luftfahrthindernis
>
> -> issue exists already without mouse icon change: if you click on the top
> north part of the single obstacle , there is a loading icon showing up but
> nothing happens. so as well here: show the mouse over icon change based on
> the highlight geometry
>
> -> IMHO: if we use in general the highlight geometry as base for the mouse
> icon change usability would be nice. (but this will have a serious load on
> the back-end I assume)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2172#issuecomment-77683159>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://t.signaletre.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XYgf6M1WN1qfVVMRYFRqW8qlQyH56dByMf1q8vLz02?t=http%3A%2F%2Ftwitter.com%2Fcedricmoullet&si=6057645868318720&pi=0f6712c7-0672-4abe-a1a2-e2e05a5f2edc>

**Comment by gjn** [comment]
@davidoesch @cedricmoullet  We need to act soon. This feature is in master and we'll deploy tomorrow/Thursday.

**Comment by davidoesch** [comment]
Apply it

**Comment by cedricmoullet** [comment]
+1

On Tue, Mar 10, 2015 at 7:44 AM, davidoesch <notifications@github.com>
wrote:

> Apply it
>
> On Tue, Mar 10, 2015 at 7:36 AM, Gilbert Jeiziner
> <notifications@github.com> wrote:
>
> > @davidoesch @cedricmoullet We need to act soon. This feature is in
> master and we'll deploy tomorrow/Thursday.
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/geoadmin/mf-geoadmin3/pull/2172#issuecomment-78002937
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2172#issuecomment-78003580>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://t.signaletre.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XYgf6M1WN1qfVVMRYFRqW8qlQyH56dByMf1q8vLz02?t=http%3A%2F%2Ftwitter.com%2Fcedricmoullet&si=6057645868318720&pi=b296f76a-ed6c-496b-fd12-4fbf7fa18244>

**Comment by gjn** [comment]
So we deploy current master?

**Comment by cedricmoullet** [comment]
yes

On Tue, Mar 10, 2015 at 8:14 AM, Gilbert Jeiziner <notifications@github.com>
wrote:

> So we deploy current master?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/geoadmin/mf-geoadmin3/pull/2172#issuecomment-78005938>
> .
>



-- 
Twitter: http://twitter.com/cedricmoullet
Linked In: http://www.linkedin.com/in/cedricmoullet

<http://t.signaletre.com/e1t/c/5/f18dQhb0S7lC8dDMPbW2n0x6l2B9nMJN7t5XYgf6M1WN1qfVVMRYFRqW8qlQyH56dByMf1q8vLz02?t=http%3A%2F%2Ftwitter.com%2Fcedricmoullet&si=6057645868318720&pi=e24ca6e1-c6ee-48dd-d8aa-a3dc86e0ea46>
